torch/onnx/_backend/core.py

```diff
@@ -0,0 +1,852 @@
+                args,
+                input_devices,
+                output_names,
+                (prim_outputs,),
+                output_devices,
+                self.preallocate_output,
+            )
+            assert len(onnx_outputs) == 1
+            if self._assert_allclose_to_baseline:
+                # Compute baseline.
+                baseline_outputs = torch._prims.executor.execute(
+                    graph_module, *args, executor="aten"
+                )
+                # Ensure output tensor is close to the corresponding baseline.
+                _assert_allclose_with_detailed_error_message(
+                    onnx_outputs[0], baseline_outputs
+                )
+            return onnx_outputs[0]
+
+    def compile(self, graph_module: torch.fx.GraphModule, args) -> torch.fx.GraphModule:
+        # FX graph based partitioning based on ONNX supported ops.
+        if graph_module in self._partitioner_cache:
+            partitioned_prim_graph_module = self._partitioner_cache[graph_module]
+        else:
+            prim_graph_module = graph_module
+            # TODO(wschin): this is required for removing aten::_to_copy in _replace_to_copy_with_to.
+            _replace_to_copy_with_to(prim_graph_module)
+            partitioner = CapabilityBasedPartitioner(
+                prim_graph_module,
+                self._supported_ops,
+                allows_single_node_partition=True,
+            )
+            partitioned_prim_graph_module = partitioner.partition_and_fuse()
+            self._partitioner_cache[graph_module] = partitioned_prim_graph_module
+
+            # Overriding fused_module's __call__() function with ort_acclerated_call()
+            # This loop goes through all graph partitions (each of them is an ONNX-representable graph)
+            # and override their _wrappped_call function with _ort_accelerated_call.
+            # Inside _ort_accelerated_call, the partition's graph is exported into ONNX and executed by ORT.
+            for node in partitioned_prim_graph_module.graph.nodes:
+                # TODO(wschin): use a better way to identify fused submodule
+                if node.op == "call_module" and "fused_" in node.name:
+                    fused_module = getattr(partitioned_prim_graph_module, node.name)
+                    # self.ort_acclerated_call is responsible for exporting graph to ONNX,
+                    # creating ORT session, and running ORT session.
+                    fused_module._wrapped_call = self._ort_acclerated_call
+
+        return partitioned_prim_graph_module
+
+    def __call__(
+        self, graph_module: torch.fx.GraphModule, args
+    ) -> torch.fx.GraphModule:
+        return self.compile(graph_module, args)
+
+
+def make_aot_ort(dynamic: bool = True):
+    """Wrap OrtBackend as PyTorch's AOT compiler.
+
+    Example usages:
+        import torch
+        from onnxruntime.training.torchdynamo.register_backend import make_aot_ort
```

Comment:
Example needs updating to reflect this PR. Would be good if the examples can be shown using the torch.compile decorator?

Reply:
+1 plz use xdoctest style such that example code is checked by CI. https://github.com/pytorch/pytorch/blob/3a07dfde488ba974dc7ed7f8f24e9bcf64ada053/torch/onnx/_internal/exporter.py#L413-L437

Reply:
Here is the new example.
```
         # xdoctest: +REQUIRES(env:TORCH_DOCTEST_ONNX) 
         >>> import copy
         >>> import torch 
         >>> from torch.onnx._backend.core import make_aot_ort
         >>> class MyModel(torch.nn.Module):  # Dummy model 
         ... def __init__(self) -> None: 
         ...     super().__init__() 
         ...     self.linear = torch.nn.Linear(2, 2) 
         ... def forward(self, x): 
         ...     out = self.linear(x) 
         ...     return out 
         >>> model = MyModel()
         >>> aot_ort, _ = make_aot_ort(dynamic = True)
         >>> compiled_model = torch.compile(copy.deepcopy(model), backend=aot_ort, dynamic=True)
         >>> x = torch.randn(2, 2, 2)
         >>> y_baseline = model(x)
         >>> y_compiled = compiled_model(x)
         >>> torch.testing.assert_close(y_baseline, y_compiled)
```